docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,272 @@
+
+```
+(RANDPRIME) \forall P:Int . isPrime(P) -> (randPrime() -> * P)
+// equivalent form(s)
+//          \forall P:Int . (randPrime() /\ isPrime(P)) -> *P
+//          \forall P:Int . randPrime() -> (isPrime(P) -> * P)
+//          randPrime() -> \forall P:Int . (isPrime(P) -> * P)
+//          remove \forall P:Int in all above three
+```
+
+where `isPrime(_)` is a predicate that can be defined in the usual way.
+
+As another example, `randDifferentFromList(Is)` takes a list `Is` of integers
+and can rewrite in one step to any integer that is not in `Is`.
+Here is its axiom
+
+```
+(RANDLIST) \forall Is:List{Int} . \forall I:Int . \neg(I \in Is) -> randDifferentFromList(Is) -> * I
+```
+
+### Frontend Syntax
+
+Since all variables are universally quantified, `rand()` and its variants could
+be written in the K frontend using regular rules and variables:
+
+```
+syntax Exp ::= rand ()
+rule rand() => X:Int
+
+syntax Exp ::= randPrime ()
+rule randPrime() => X:Int
+  requires isPrime(X)
+
+syntax Exp ::= randList (List{Int})
+rule randList(Is) => I:Int
+  requires notBool (I inList{Int} Is)
+```
+
+Note 1: all above are not function symbols, but languages constructs.
+
+Note 2: Currently the frontend does not allow rules with universally quantified
+variables in the RHS which are not bound in the LHS.
+
+## Example B: Fresh Integer Construct `fresh(Is)`
+
+The fresh integer construct `fresh(Is)` is a language construct and part of the
+PL syntax.
+
+```
+Exp ::= ... | "fresh" "(" List{Int} ")"
+```
+
+### Axioms
+
+The intended semantics of `fresh(Is)` is that it can always rewrite to an integer
+that in not in `Is`.
+
+Note that `fresh(Is)` and `randDifferentFromList(Is)` are different; the former
+does not _need_ to be able to rewrite to every integers not in `Is`,
+while the latter requires so.
+
+For example, it is _correct_ to implement `fresh(Is)` so it always returns the
+smallest positive integer that is not in `Is`, but same implementation for
+`randDifferentFromList(Is)` is considered _incorrect_.
+In other words, there exist multiple correct implementations of `fresh(Is)`,
+some of which may be deterministic, but there only exists a unique
+implementation of `randDifferentFromList(Is)`.
+Finally, note that `randDifferentFromList(Is)` is a _correct_ implementation
+for `fresh(Is)`.
+
+Here is the axiom of `fresh(Is)`
+
+```
+(FRESH) \forall Is:List{Int} . \exists I:Int . \neg( I \in Is) /\ (fresh(Is) -> * I)
+// equivalent form(s)
+//      \forall Is:List{Int} . fresh(Is) -> \exists I:Int . \neg( I \in Is) /\ * I
+//      \forall Is:List{Int} . fresh(Is) -> \exists I:Int . * (\neg( I \in Is) /\ I)
+//      \forall Is:List{Int} . fresh(Is) -> * \exists I:Int . \neg( I \in Is) /\ I
+```
+
+### Frontend Syntax
+
+We use the following K syntax to define `fresh(Is)`
+
+```
+syntax Exp ::= fresh (List{Int})
+rule fresh(Is:List{Int}) => ?I:Int
+  ensures notBool (?I inList{Int} Is)
+```
+
```

Comment:
LGTM, too. So the idea is that the ?-variables are existentially quantified after the universally quantified ones, also at the top. Interestingly, you can have both ?- and non-? fresh variables in the RHS of a rule, the former being existentially quantified after the latter (so they ? variables depend, a la Skolem, on the non-? ones; it may be a good idea to have such an example, too).

Reply:
We could, but this again looks more like a case in which it might be clearer to have a proper ML rule.  I would not add an example for this unless it's a compelling and useful one, which I can't thing of at this time.